### ISSUE TITLE
IotHubUnauthorizedAccess on device provisioning

### ISSUE THREAD

**Issue by MattHoneycutt** [Azure/azure-iot-sdk-csharp#898]
- **OS, version, SKU and CPU architecture used:** Windows 10 Desktop x64
- **Application's .NET Target Framework :** netcoreapp2.1
- **Device:** Desktop
- **SDK version used:** 

Microsoft.Azure.Devices.Client - 1.19.0
Microsoft.Azure.Devices.Provisioning.Client - 1.2.2
Microsoft.Azure.Devices.Provisioning.Transport.Amqp - 1.1.5

## Description of the issue:
I'm an instructor for Pluralsight, and I've got a sample app that works fine on some machines, with some IoT Hubs, but gets the following error on some hubs and DPS instances:

```
Unhandled Exception: Microsoft.Azure.Devices.Provisioning.Client.ProvisioningTransportException: AMQP transport exception ---> Microsoft.Azure.Amqp.AmqpException: {"errorCode":401002,"trackingId":"c93b0cd9-c042-48f1-8e22-e7654a6eec5e","message":"IotHubUnauthorizedAccess","timestampUtc":"2019-01-26T23:12:21.4244905Z"}
at Microsoft.Azure.Amqp.AsyncResult.End[TAsyncResult](IAsyncResult result)
at Microsoft.Azure.Amqp.AmqpObject.OpenAsyncResult.End(IAsyncResult result)
at Microsoft.Azure.Amqp.AmqpObject.EndOpen(IAsyncResult result)
at System.Threading.Tasks.TaskFactory`1.FromAsyncCoreLogic(IAsyncResult iar, Func`2 endFunction, Action`1 endAction, Task`1 promise, Boolean requiresSynchronization)
--- End of stack trace from previous location where exception was thrown ---
at Microsoft.Azure.Devices.Provisioning.Client.Transport.AmqpClientConnection.OpenAsync(TimeSpan timeout, Boolean useWebSocket, X509Certificate2 clientCert, IWebProxy proxy)
at Microsoft.Azure.Devices.Provisioning.Client.Transport.ProvisioningTransportHandlerAmqp.RegisterAsync(ProvisioningTransportRegisterMessage message, CancellationToken cancellationToken)
--- End of inner exception stack trace ---
at Microsoft.Azure.Devices.Provisioning.Client.Transport.ProvisioningTransportHandlerAmqp.RegisterAsync(ProvisioningTransportRegisterMessage message, CancellationToken cancellationToken)
at DeviceProvisioningSample.Device.Provision() in c:\Repos\iothubdemos\ps-create-iot-solutions\device-provisioning-sample\Device.cs:line 34
at DeviceProvisioningSample.Program.Main(String[] args) in c:\Repos\iothubdemos\ps-create-iot-solutions\device-provisioning-sample\Program.cs:line 26
at DeviceProvisioningSample.Program.<main>(String[] args)
```

Again, this works fine for some students, but not others.  I've had no luck finding a common culprit.   We've tried deleting and recreating hubs and DPS instances, but it doesn't seem to help.  

## Code sample exhibiting the issue:
https://github.com/MattHoneycutt/ps-create-iot-solutions

## Console log of the issue:
n/a

Any suggestions on what I can try to troubleshoot or debug this further?

**Comment by prmathur-microsoft** [comment]
@alextolp Could you please assist with this issue? 
"errorCode":401002,"trackingId":"c93b0cd9-c042-48f1-8e22-e7654a6eec5e","message":"IotHubUnauthorizedAccess","timestampUtc":"2019-01-26T23:12:21.4244905Z"

**Comment by alextolp** [comment]
@MattHoneycutt are you still seeing this issue? If so, do you have a recent trackingId? The trackingId in this thread already aged out of our logs.

One thing I would recommend is to enable Diagnostic logging via Diagnostic settings in the Azure Portal to get some insights about the error.

**Comment by MattHoneycutt** [comment]
@alextolp Yeah, I'm trying to get a recent trackingId.  I'll post as soon as I have one.

**Comment by MattHoneycutt** [comment]
Hi @alextolp !  Here's a more recent error, including a tracking number:

```
"Microsoft.Azure.Devices.Provisioning.Client.ProvisioningTransportException: AMQP transport exception ---> Microsoft.Azure.Amqp.AmqpException: {\"errorCode\":401002,\"trackingId\":\"8df2d992-9f94-4c36-9cfb-5711574a886f\",\"message\":\"IotHubUnauthorizedAccess\",\"timestampUtc\":\"2019-04-19T17:10:01.1692317Z\"}\r\n   at Microsoft.Azure.Amqp.AsyncResult.End[TAsyncResult](IAsyncResult result)\r\n   at Microsoft.Azure.Amqp.AmqpObject.OpenAsyncResult.End(IAsyncResult result)\r\n   at Microsoft.Azure.Amqp.AmqpObject.EndOpen(IAsyncResult result)\r\n   at System.Threading.Tasks.TaskFactory`1.FromAsyncCoreLogic(IAsyncResult iar, Func`2 endFunction, Action`1 endAction, Task`1 promise, Boolean requiresSynchronization)\r\n--- End of stack trace from previous location where exception was thrown ---\r\n   at Microsoft.Azure.Devices.Provisioning.Client.Transport.AmqpClientConnection.OpenAsync(TimeSpan timeout, Boolean useWebSocket, X509Certificate2 clientCert, IWebProxy proxy)\r\n   at Microsoft.Azure.Devices.Provisioning.C
lient.Transport.ProvisioningTransportHandlerAmqp.RegisterAsync(ProvisioningTransportRegisterMessage message, CancellationToken cancellationToken)\r\n   --- End of inner exception stack trace ---\r\n   at Microsoft.Azure.Devices.Provisioning.Client.Transport.ProvisioningTransportHandlerAmqp.RegisterAsync(ProvisioningTransportRegisterMessage message, CancellationToken cancellationToken)\r\n   at DeviceProvisioningSample.Device.Provision() in C:\\ps-create-iot-solutions-master\\device-provisioning-sample\\Device.cs:line 37"
```

Please let me know what other info I can provide.

**Comment by alextolp** [comment]
Here's what we're seeing server-side when opening the connection -

System.TimeoutException: The operation did not complete within the allocated time 00:01:00 for object in-connection184557.

@MattHoneycutt is the behavior that certain clients aren't seeing this while others are seeing it all the time? Or is it that the same client is seeing this sometimes.

Also, are you able to repro this with websockets?

**Comment by MattHoneycutt** [comment]
@alextolp It's the former: some clients are seeing this all the time, others never see it.  

I haven't tried switching the protocol... I can have them try that.  Are you thinking it could be firewall related?

**Comment by alextolp** [comment]
@MattHoneycutt yes, I suspect it's firewall related. This is currently being blocked at the connection layer, so perhaps websockets can resolve this.

**Comment by MattHoneycutt** [comment]
@alextolp I had a student try it.  He disabled his firewall (we think) and switched it to use websockets as a fallback, but he received the same error.

```
{Microsoft.Azure.Amqp.AmqpException: {"errorCode":401002,"trackingId":"adf90349-7b71-4a17-989b-8cd206b6a4ad","message":"IotHubUnauthorizedAccess","timestampUtc":"2019-04-22T20:36:17.4577525Z"}
   at Microsoft.Azure.Amqp.AsyncResult.End[TAsyncResult](IAsyncResult result)
   at Microsoft.Azure.Amqp.AmqpObject.OpenAsyncResult.End(IAsyncResult result)
   at Microsoft.Azure.Amqp.AmqpObject.EndOpen(IAsyncResult result)
   at System.Threading.Tasks.TaskFactory`1.FromAsyncCoreLogic(IAsyncResult iar, Func`2 endFunction, Action`1 endAction, Task`1 promise, Boolean requiresSynchronization)
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.Azure.Devices.Provisioning.Client.Transport.AmqpClientConnection.OpenAsync(TimeSpan timeout, Boolean useWebSocket, X509Certificate2 clientCert, IWebProxy proxy)
   at Microsoft.Azure.Devices.Provisioning.Client.Transport.ProvisioningTransportHandlerAmqp.RegisterAsync(ProvisioningTransportRegisterMessage message, CancellationToken cancellationToken)}
```

**Comment by alextolp** [comment]
Seeing the same issue but with WS now.... still looking like something related to connectivity.

Does it work with HTTP or MQTT?

**Comment by MattHoneycutt** [comment]
Still waiting to hear back on whether switching to HTTP or MQTT helped, but here's a stack trace and recent tracking ID from another student:

```
Microsoft.Azure.Devices.Provisioning.Client.ProvisioningTransportException: AMQP transport exception ---> Microsoft.Azure.Amqp.AmqpException: {"errorCode":401002,"trackingId":"3df85891-1dc0-4378-a6d7-2facc8d91063","message":"IotHubUnauthorizedAccess","timestampUtc":"2019-04-23T15:50:48.9934613Z"}
at Microsoft.Azure.Amqp.AsyncResult.End[TAsyncResult](IAsyncResult result)
at Microsoft.Azure.Amqp.AmqpObject.OpenAsyncResult.End(IAsyncResult result)
at Microsoft.Azure.Amqp.AmqpObject.EndOpen(IAsyncResult result)
at System.Threading.Tasks.TaskFactory`1.FromAsyncCoreLogic(IAsyncResult iar, Func`2 endFunction, Action`1 endAction, Task`1 promise,
Boolean requiresSynchronization)
--- End of stack trace from previous location where exception w as thrown ---
at Microsoft.Azure.Devices.Provisioning.Client.Transport.AmqpClientConnection.OpenAsync(TimeSpan timeout, Boolean useWebSocket, X509Certificate2 clientCert, IWebProxy proxy)
at Microsoft.Azure.Devices.Provisioning.Client.Transport.ProvisioningTransportHandlerAmqp.RegisterAsync(ProvisioningTransportRegisterMessage message, CancellationToken cancellationToken)
--- End of inner exception stack trace ---
at Microsoft.Azure.Devices.Provisioning.Client.Transport.ProvisioningTransportHandlerAmqp.RegisterAsync(ProvisioningTransportRegisterMessage message, CancellationToken cancellationToken)
at DeviceProvisioningSample.Device.Provision() in C:\Users\User\Documents\Coding C#\device-provisioning-sample\Device.cs:line 33
at DeviceProvisioningSample.Program.Main(String[] args) in C:\User\User\Documents\Coding C#\device-provisioning-sample\Program.cs:line 26
at DeviceProvisioningSample.Program.<main>(String[] args)
```

**Comment by MattHoneycutt** [comment]
Hi @alextolp,

We're still getting the same issue with HTTP: 

```
at Microsoft.Azure.Devices.Provisioning.Client.Transport.ProvisioningTransportHandlerHttp.RegisterAsync(ProvisioningTransportRegisterMessage message, CancellationToken cancellationToken)
   at DeviceProvisioningSample.Device.Provision() in C:\Users\VertexLt\Downloads\ps-create-iot-solutions-master\device-provisioning-sample\Device.cs:line 37
{"errorCode":401002,"trackingId":"a882eee2-be50-4188-90c6-32d9506cb85a","message":"Unauthorized","timestampUtc":"2019-04-25T13:18:41.9121071Z"}
```

Is it the same issue on the backend? 

Thanks!

**Comment by alextolp** [comment]
@MattHoneycutt  this one is slightly different. What auth is being used? We don't see any credential (authorization header or x509 cert) on our end.

**Comment by Cryan24** [comment]
Sorry to pile on, I am have the same problem, using a X509 leaf cert with an enrollment group configured(with CA cert).

Microsoft.Azure.Devices.Provisioning.Client.ProvisioningTransportException: '{"errorCode":401002,"trackingId":"cda45c50-d852-4c71-ae23-cbd24844b604","message":"Unauthorized","timestampUtc":"2019-05-02T11:33:18.7971239Z"}'

**Comment by alextolp** [comment]
@Cryan24 also seeing an issue where the X509 certificate is missing in the connection. How are you initializing the SecurityProvider?

**Comment by Cryan24** [comment]
@alextolp like this:

` private const string GlobalDeviceEndpoint = "global.azure-devices-provisioning.net";
        public static int Main(string[] args)
        {	   
			var cert = new X509Certificate2("myCert.pem", "certPass");	  
			using (var security = new SecurityProviderX509Certificate(cert))

            // Select one of the available transports:
            // To optimize for size, reference only the protocols used by your application.
           // using (var transport = new ProvisioningTransportHandlerAmqp(TransportFallbackType.TcpOnly))
             using (var transport = new ProvisioningTransportHandlerHttp())
            // using (var transport = new ProvisioningTransportHandlerMqtt(TransportFallbackType.TcpOnly))
            // using (var transport = new ProvisioningTransportHandlerMqtt(TransportFallbackType.WebSocketOnly))
            {
                ProvisioningDeviceClient provClient =
                    ProvisioningDeviceClient.Create(GlobalDeviceEndpoint, "[myDPSIdScope]", security, transport);

                var sample = new ProvisioningDeviceClientSample(provClient, security);
                sample.RunSampleAsync().GetAwaiter().GetResult();
            }

            return 0;
        }`

I managed to get it working by converting my cert from .pem to .pfx in openssl, so I am no longer stuck for a fix but hopefully this gives you something to trace for others who are using .pem cert files.

**Comment by MattHoneycutt** [comment]
> @MattHoneycutt this one is slightly different. What auth is being used? We don't see any credential (authorization header or x509 cert) on our end  .

@alextolp  We're using X509 certs. The cert is a PFX file.  Here's the relevant code:

```
            using (var securityProvider = new SecurityProviderX509Certificate(certificate))
            using (var transport = new ProvisioningTransportHandlerAmqp(TransportFallbackType.TcpWithWebSocketFallback))
            {
                var client = ProvisioningDeviceClient.Create("global.azure-devices-provisioning.net", this.scopeId, securityProvider, transport);

```

**Comment by MattHoneycutt** [comment]
@alextolp Any update on this?  Are we doing anything wrong in our code for creating/using the provisioning client? 

Thanks!

**Comment by alextolp** [comment]
@MattHoneycutt  can you please share your:

1) OS and build
2) .NET version

cc: @prmathur-microsoft

**Comment by Shreyasnie** [comment]
i am also getting same error even after using pfx file

"Service Error: 401002 - CA certificate not found (TrackingID: 51ef4a96-6c05-4a3c-8694-b657f38e5dea Time: 2019-05-14T05:20:00.5935533Z"

**Comment by alextolp** [comment]
@Shreyasnie your issue appears to be caused by device presenting a certificate that doesn't match any CA registered with DPS. Please double check that the CA is registered and that the device presents the full cert chain when establishing the TLS connection

**Comment by Shreyasnie** [comment]
@alextolp following are the steps which i executed

1)Create CA certificate. 
2)Uploaded CA certificate to certificates section under DPS. 
3)Generated verification code. 
4)Generated verification certificate using verification code as common name
5)uploaded verification certificate and verified successfully.
6)created group enrollment with Root certificate.
7)generated device specif pfx.
tried to provision device using device specific certificate it throws Service Error 401002 Certificate not found

Please let me know if i am missing some steps

**Comment by MattHoneycutt** [comment]
@alextolp This particular student is using:

Microsoft Windows 10 Pro 10.0.17763 N/A Build 17763
.NET Core 2.2

**Comment by alextolp** [comment]
@MattHoneycutt  what NuGet package version are you using?

**Comment by alextolp** [comment]
@Shreyasnie from the logs for the trackingId you've shared it appears that there's an Intermediate CA between your Device cert and your Root CA.

This intermediate CA should be presented in TLS so we can build the entire chain on the server side.

**Comment by MattHoneycutt** [comment]
@alextolp Relevant versions:

Microsoft.Azure.Devices.Client: 1.19.0
Microsoft.Azure.Devices.Provisioning.Client:  1.2.2
Microsoft.Azure.Devices.Provisioning.Transport.Amqp: 1.1.5

**Comment by abhipsaMisra** [comment]
@MattHoneycutt It looks like an issue with the process of generating the cert/ adding the enrollment into DPS. I followed the following steps and was able to run the provided sample:

1. I used our [certificate generator](https://github.com/Azure-Samples/azure-iot-samples-csharp/blob/master/provisioning/Samples/device/X509Sample/GenerateTestCertificate.ps1) to generate the X509 certificate. It creates a self-signed certificate and stores it in `"Cert:\CurrentUser\My"`. The exported pfx is password protected as well.
1. I created an individual enrollment record in DPS, with the generated public key file (.cer) as the primary certificate.
1. I updated the X509 certificate [import](https://github.com/MattHoneycutt/ps-create-iot-solutions/blob/master/device-provisioning-sample/Device.cs#L56) code to take in a password as well, and ran the sample. 
`certificateCollection.Import(key, certificatePassword, X509KeyStorageFlags.UserKeySet);`

The device was successfully provisioned to my IoTHub, sent D2C message and then closed the connection.

**Comment by prmathur-microsoft** [comment]
@MattHoneycutt As per the above comment, can you describe how are you generating certs?

**Comment by MattHoneycutt** [comment]
@prmathur-microsoft @abhipsaMisra  I'm using .NET Core to generate the cert:

https://github.com/MattHoneycutt/ps-create-iot-solutions/blob/master/device-provisioning-sample/CertificateFactory.cs

The certs generated by this seem to work fine on most machines, but on some, the certs enroll fine, but get the error described above.

**Comment by prmathur-microsoft** [comment]
Can you map your changes to https://github.com/Azure-Samples/azure-iot-samples-csharp/blob/master/provisioning/Samples/device/X509Sample/GenerateTestCertificate.ps1 as we verified our certs work on that and let us know if you need anything more than what is given in the scripts.

**Comment by PawelFiderek** [comment]
Same here, tracking id: 2c8aa133-84b6-44f6-8e6c-d7369b63e0ca
DPS is configured with root cert and the cert is verified. Enrollment group is also ready.

**Comment by MattHoneycutt** [comment]
@prmathur-microsoft The two are fairly different since one is using a PowerShell cmdlet and the other the .NET API, but a few changes jump out at me:

1) The PowerShell version exports the cert in ASCII format.  My .NET version uses the default for File.WriteAllText, which is UTF-8.
2) I'm not setting a password for the exported cert.

Should either of those differences matter?

That said, I do want to point out again that my .NET version works fine for most people.  For the people that have issues, they just get the original error mentioned above during the provisioning process itself.  Azure portal seems to take in the public certificate just fine, and the .NET client code is (seemingly) able to load the private cert just fine as well.

**Comment by alextolp** [comment]
There are quite a few different issues in here and I just want to clarify what we've seen so far:

With X509 enrollment groups specifically we've seen that the cert presented by the device in TLS isn't signed by a CA registered in DPS. Either proof of possession wasn't completed, or the CA registered with DPS has a subject name that matches the issuer on the device certificate but a different key was used to sign the device certificate. 

We've also seen cases where the certificate chain looks something like:

  Root -> CA1 -> CA2 -> Device1

CA1 is registered with DPS, but the Device does not present CA2 in the TLS and there's no way to build the chain to CA1.

Additionally, we've seen cases where a Device certificate is passed to the SDK but we don't see mutual TLS happen and don't see any credential server-side. @MattHoneycutt - it seems like this is the case with some of the trackingIds you've shared. Any way you can capture a network trace and confirm whether mTLS occurs? This can be done with something like wireshark.

**Comment by MattHoneycutt** [comment]
@alextolp Sure, I'll try to get that for you.

If mTLS *isn't* being established... what would cause that?  Could that be a config issue with the code, or something deeper? 

Thanks!

**Comment by alextolp** [comment]
@MattHoneycutt really unclear at this point... likely some bug we haven't been able to track down, and wireshark can at least confirm whether it's client-side or server-side.

**Comment by MattHoneycutt** [comment]
@alextop I have a Wireshark dump from one of my students that is impacted by this issue.  I see TLS 1.2 traffic, but I'm honestly not sure what I'm looking for to determine whether an mTLS session was established... any pointers?

**Comment by PawelFiderek** [comment]
I think i know the issue with IoT hub non authorized exception. Cert i was using had ni private key incloded. After using open ssl to generate one (compile) with key inside everything start to work without any code change... i think it is worth to include this info as a dps api limitation/restriction :)

**Comment by alextolp** [comment]
Thanks for your insights @PawelFiderek !

@MattHoneycutt  can you check the "HasPrivateKey" flag on the client certificate for folks who are able to repro the issue? It must be true in order for mTLS to succeed.

https://docs.microsoft.com/en-us/dotnet/api/system.security.cryptography.x509certificates.x509certificate2.hasprivatekey?view=netframework-4.8

**Comment by MattHoneycutt** [comment]
@alextolp The certificates do indeed have "HasPrivateKey = true."  In fact, I can take one of the certificates my students generate and use it on my machine with DPS just fine.

**Comment by MattHoneycutt** [comment]
@alextolp Any update here?  I think we've confirmed that the issue doesn't have anything to do with "HasPrivateKey" since the certificates in question do have a private key, and they do work as-is if we transfer them to another machine.  I do have a Wireshark dump, I'm just not sure what I'm looking for in it.

Thanks for all your help so far, hopefully we're getting close to the bottom of this. :)

**Comment by MattHoneycutt** [comment]
@alextolp Any guidance here?  I'm happy to dig through the Wireshark dump, or upload it here, but I need some direction on what I'm looking for.  

Thanks!

**Comment by BhupalPatil** [comment]
> @alextolp following are the steps which i executed
> 1)Create CA certificate. 
> 2)Uploaded CA certificate to certificates section under DPS. 
> 3)Generated verification code. 
> 4)Generated verification certificate using verification code as common name
> 5)uploaded verification certificate and verified successfully.
> 6)created group enrollment with Root certificate.
> 7)generated intermediate certificate pfx.
> 8)generated device specif pfx using intermediate certificate.
> tried to provision device using device specific certificate it throws Service Error 401002 Certificate not found
> Please let me know if i am missing some steps
> I also checked device certificates it presents all 3 certificates (Root.cer,Intermediate.pfx and Device.pfx)
> following code from github was used to create certificates
> https://github.com/rwatjen/AzureIoTDPSCertificates

@Shreyasnie,
were you able to get that issue sorted? I'm facing same issue followed exactly same steps your mentioned above.

**Comment by mparkitny** [comment]
I had the exact same issue.
Turned out the problem was with incorrect version of Windows. Once I tried those steps on a Windows 10 Pro machine instead of Home, everything went smoothly.
It would be nice if error messages were a bit less misleading.

**Comment by mparkitny** [comment]
@MattHoneycutt 
The original one. 
I tried to register to DPS from sereval different machines. On all machines with Windows Home I get Microsoft.Azure.Devices.Provisioning.Client.ProvisioningTransportException, on machines with WIndows Pro everything works perfectly. So far I haven't found another reason/difference.

**Comment by alextolp** [comment]
Thanks for your feedback @mparkitny ! I haven't had a chance to test on Windows 10 Home yet, but I'd like to add some comments from my observations. Most of the feedback here is around two-tier PKIs, e.g.

rootCA -> issuingCA -> deviceCert

With this setup there are multiple options, but I'll cover the following two:

1) Go through proof of possession on the issuingCA, create a group for the issuingCA, and provision devices using the deviceCert. This is the recommended option, and should work without any tinkering as long as the CA cert is in the "Verified" state and a group exists and is enabled for provisioning.

2) Go through proof of possession on the rootCA, create a group for the rootCA, and provision devices using the deviceCert. Root CA is meant be kept offline, so this option is not recommended in production, but I understand that it can be useful for certain dev scenarios. 

For option2, the issuing CA must be present in the TLS handshake for the server to build and validate the chain from the deviceCert to the root CA. On Windows, intermediate CAs are required to be installed in the cert store for them to appear in the TLS handshake, and there's a utility class to make it work - 

https://github.com/Azure/azure-iot-sdk-csharp/blob/c93b167058708aead0d2849972c21dabf815b945/provisioning/device/src/CertificateInstaller.cs#L37

In order to invoke the utility, you must initialize the Security Provider with the issuingCA present in the certificateChain collection.

https://github.com/Azure/azure-iot-sdk-csharp/blob/c93b167058708aead0d2849972c21dabf815b945/shared/src/SecurityProviderX509Certificate.cs#L23

You can use WireShark to validate the client certificates are present in the handshake. Here's what worked for me:

1) Start capturing WireShark trace on your Ethernet adapter
2) Run through the provisioning code and let WireShark collect the traces of the entire provisioning process
3) Stop capturing WireShark traces
4) Find the IP address of the global device endpoint:
    >ping dpsci.azure-devices-provisioning-int.net
5) Filter events from / to the global device endpoint, e.g:
    >tls and (ip.dst == 40.113.233.146 or ip.src == 40.113.233.146)
6) Locate and click the Certificate, Client Key Exchange, etc. event. It will be the one immediately following Server Hello Done event
![image](https://user-images.githubusercontent.com/12418633/63994785-b6b3d280-caaa-11e9-810e-037e8a5d660f.png)
7) Locate the Certificate section of the TLS Handshake protocol
![image](https://user-images.githubusercontent.com/12418633/63994978-7e60c400-caab-11e9-9bbc-56a2c017a188.png)

In my example, I used a two-tier PKI, that looks like:

root-59ecbc49-caf7-49dc-a143-af49175cd955 -> intermediate-59ecbc49-caf7-49dc-a143-af49175cd955 -> device-59ecbc49-caf7-49dc-a143-af49175cd955-0

Which you can see in the screen cap. 

Please double check that you're using the correct constructor of SecurityProviderX509Certificate and that you see the issuingCA in the TLS handshake.

**Comment by alextolp** [comment]
One other thing - perhaps there are permission issues with installing certs into CurrentUser\CA store. @MattHoneycutt  @mparkitny can you try installing your rootCA and issuingCA to CurrentUser\My store prior to invoking the provisioning API. You can use the following utility if needed:

        public static void InstallCertificates(IEnumerable<X509Certificate2> certs)
        {
            X509Store store = new X509Store(StoreName.My, StoreLocation.CurrentUser);
            try
            {
                store.Open(OpenFlags.ReadWrite);

                foreach (var cert in certs)
                {
                    store.Add(cert);
                }
            }
            finally
            {
                store.Close();
            }
        }

**Comment by ppparihar** [comment]
I'm facing the same issue.
`Unhandled Exception: Microsoft.Azure.Devices.Provisioning.Client.ProvisioningTransportException: AMQP transport exception ---> Microsoft.Azure.Amqp.AmqpException: {"errorCode":401002,"trackingId":"cc04cc2b-857a-4261-a2fd-4e10266ec600","message":"IotHubUnauthorizedAccess","timestampUtc":"2019-10-04T06:27:12.7837167Z"}`

**Comment by alextolp** [comment]
@ppparihar  have you tried the troubleshooting steps above?

**Comment by MattHoneycutt** [comment]
@alextolp Installing the certs in the local store first does not work.  Same error as before, even with the latest NuGet packages.  Here's the latest error with a tracking ID:

```
Unhandled Exception: Microsoft.Azure.Devices.Provisioning.Client.ProvisioningTransportException: AMQP transport exception ---> Microsoft.Azure.Amqp.AmqpException: {"errorCode":401002,"trackingId":"7fa9f4ba-5ddd-48a0-aa54-a9f0be826f34","message":"IotHubUnauthorizedAccess","timestampUtc":"2019-10-11T04:34:47.5937293Z"}
```

**Comment by alextolp** [comment]
@MattHoneycutt can you collect WireShark traces and check what client certs are sent over the wire?

**Comment by MattHoneycutt** [comment]
Hi @SRIVASTAVAABHISHEKGAURAV, can you please try the latest version of the code in this repo: 
https://github.com/matthoneycutt/ps-create-iot-solutions

**Comment by egitura** [comment]
@MattHoneycutt Thanks for the support, I had the same issue and it seems the Pluralsight exercise files are still out of date. With your most recent version from github it works fine.

**Comment by MattHoneycutt** [comment]
@egitura Awesome, thanks for letting me know the github version works.  I'm in the process of submitting updates to Pluralsight now.

**Comment by SRIVASTAVAABHISHEKGAURAV** [comment]
Hello @MattHoneycutt  Thank you for the updated version.

 Unfortunately I am facing "Unhandled Exception: System.InvalidOperationException: Something went wrong while trying to provision". issue. when executing command (dotnet run {id scope}). 
Attaching herewith the screenshots of my run. Let me know if I am missing something here.

cc: @egitura Could you please guide me here, How did it work for you? I can try your approach.


<img width="665" alt="Windows_System32_cmd" src="https://user-images.githubusercontent.com/22419505/69170702-8e26fc00-0b20-11ea-9ede-492c273507b1.png">
<img width="570" alt="globomantics-dev-1574183267 - Microsoft Azure" src="https://user-images.githubusercontent.com/22419505/69170703-8ebf9280-0b20-11ea-8c47-ac81d07f7e59.png">

**Comment by SRIVASTAVAABHISHEKGAURAV** [comment]
Hello @MattHoneycutt Thank you for the updated version.

Unfortunately I am facing "**_Unhandled Exception: System.InvalidOperationException: Something went wrong while trying to provision__**". issue. when executing command (dotnet run {id scope}).
Attaching herewith the screenshots of my run. Let me know if I am missing something here. 
(running on windows 10 and VS2019)


cc: @egitura Could you guide me here. How did it work for you? I can try with that approach.



<img width="665" alt="Windows_System32_cmd" src="https://user-images.githubusercontent.com/22419505/69172192-65543600-0b23-11ea-932d-299037c3e679.png">
<img width="570" alt="globomantics-dev-1574183267 - Microsoft Azure" src="https://user-images.githubusercontent.com/22419505/69172193-65eccc80-0b23-11ea-8650-a188e1de4f25.png">

**Comment by egitura** [comment]
@SRIVASTAVAABHISHEKGAURAV Afraid I don't have anything to add here as I didn't do anything special, just followed the process as explained in Matt's lesson again but using the updated code. Perhaps debug and check what status you got back and see from there? Did you link the IoT hub? also did you get the updated code from github? that error is thrown on line 39, not 42 as it says in your screenshot..

**Comment by gpsagar123** [comment]
+1
Facing the same problem as mentioned above, tried with exercise file from Pluralsight and the GIT hub link https://github.com/MattHoneycutt/ps-create-iot-solutions/tree/master/device-provisioning-sample - still the problem persist
@MattHoneycutt : Can you please look into and help here.

**Comment by mjmastro** [comment]
I had trouble with this step as well. I am using PowerShell ISE on a Windows 10 Enterprise System. I ran the command: dotnet run One000D265F and it resulted in the below exception. (I'm not a .NET developer. I am a data engineer, so my experience in .NET is very limited). 

Any advice would be good. I'll continue with the course without this step completed.

Microsoft.Azure.Devices.Provisioning.Client.ProvisioningTransportException: AMQP transport exception ---> Microsoft.Azure.Amqp.AmqpException: 
{"errorCode":401002,"trackingId":"72781f42-020d-404b-a81c-de288393aef2","message":"IotHubUnauthorizedAccess","timestampUtc":"2020-03-08T00:04:35.4116335Z"}
   at Microsoft.Azure.Amqp.AsyncResult.End[TAsyncResult](IAsyncResult result)
   at Microsoft.Azure.Amqp.AmqpObject.OpenAsyncResult.End(IAsyncResult result)
   at Microsoft.Azure.Amqp.AmqpObject.EndOpen(IAsyncResult result)
   at System.Threading.Tasks.TaskFactory`1.FromAsyncCoreLogic(IAsyncResult iar, Func`2 endFunction, Action`1 endAction, Task`1 promise, Boolean requiresSynchronization)
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.Azure.Devices.Provisioning.Client.Transport.AmqpClientConnection.OpenAsync(TimeSpan timeout, Boolean useWebSocket, X509Certificate2 clientCert, 
IWebProxy proxy)
   at Microsoft.Azure.Devices.Provisioning.Client.Transport.ProvisioningTransportHandlerAmqp.RegisterAsync(ProvisioningTransportRegisterMessage message, 
CancellationToken cancellationToken)
   --- End of inner exception stack trace ---
   at Microsoft.Azure.Devices.Provisioning.Client.Transport.ProvisioningTransportHandlerAmqp.RegisterAsync(ProvisioningTransportRegisterMessage message, 
CancellationToken cancellationToken)
   at DeviceProvisioningSample.Device.Provision() in C:\myTraining\PluralSight\DP200\iot-hub\04\device-provisioning-sample\Device.cs:line 33
   at DeviceProvisioningSample.Program.Main(String[] args) in C:\myTraining\PluralSight\DP200\iot-hub\04\device-provisioning-sample\Program.cs:line 26
   at DeviceProvisioningSample.Program.<Main>(String[] args)

**Comment by timtay-microsoft** [comment]
This thread is getting cluttered and hard to follow, but it does appear a resolution was reached, so I'm going to close this issue. If anyone is still having issues here, please open a separate thread.

**Comment by az-iot-builder-01** [comment]
@prmathur-microsoft, @MattHoneycutt, @C-M-Ryan, @Shreyasnie, @abhipsaMisra, @PawelFiderek, thank you for your contribution to our open-sourced project! Please help us improve by filling out this 2-minute [customer satisfaction survey](https://forms.office.com/Pages/ResponsePage.aspx?id=v4j5cvGGr0GRqy180BHbRx8jpG79OudPqY-BojcKjkJUMFFBQjgxNkw5ODJIRjBWTzI0Nk85R0pUQiQlQCN0PWcu)

**Comment by subatta** [comment]
One possibility is that the root cert generated may have expired, in which case the interim and leaf certs would be invalid.
